test/unit/org/apache/cassandra/metrics/DecayingEstimatedHistogramReservoirTest.java

```diff
@@ -604,6 +616,110 @@ public void testSize()
+                            Uninterruptibles.sleepUninterruptibly(ThreadLocalRandom.current().nextInt(1, 10), MILLISECONDS);
+                            histogram.update(1109);
+                        }
+                    }
+                });
+            }
+            // clock update thread
+            executors.submit(() -> {
+                while (!stop.get() && !Thread.currentThread().isInterrupted())
+                {
+                    Uninterruptibles.sleepUninterruptibly(1, TimeUnit.MILLISECONDS);
+                    // x1000 speedup so that we hit rescale interval every 30 minutes / 1000 = 1.8s
+                    clock.addMillis(1000);
+                }
+            });
+            // percentiles check thread
+            executors.submit(() -> {
+                // how many times in a row p99 was suspiciously low or P50 suspiciously high
+                int consecutiveInvalidPercentiles = 0;
+
+                // how often to check the percentiles
+                int iterationDelayMillis = 100;
+
+                for (int i = 0; i < maxTestDurationMillis / iterationDelayMillis; i++)
+                {
+                    Uninterruptibles.sleepUninterruptibly(iterationDelayMillis, MILLISECONDS);
+                    Snapshot snapshot = toSnapshot.apply(histogram);
+                    double p99 = snapshot.getValue(0.99);
+                    double p50 = snapshot.getValue(0.50);
+                    logger.info("{} p99: {}", clock.now() / 1_000_000, p99);
+                    logger.info("{} p50: {}", clock.now() / 1_000_000, p50);
+                    ByteArrayOutputStream output = new ByteArrayOutputStream();
+                    snapshot.dump(output);
+                    String decayingNonZeroBuckets = Arrays.stream(output.toString().split("\n"))
+                                                          .filter(s -> !s.equals("0"))
+                                                          .collect(Collectors.joining(","));
+                    logger.info("decaying non-zero buckets:  {}", decayingNonZeroBuckets);
+                    if (p99 < 100 || p50 > 900)
+                    {
+                        consecutiveInvalidPercentiles++;
+                        logger.warn("p50 or p99 at suspicious level p50={}, p99={}", p50, p99);
+                        if (consecutiveInvalidPercentiles > tooManySuspiciousPercentilesThreshold)
+                        {
+                            failed.set(true);
+                            stop.set(true);
+                            break;
+                        }
+                    }
+                    else
+                    {
+                        consecutiveInvalidPercentiles = 0;
+                    }
+                }
+                stop.set(true);
+            });
+            executors.shutdown();
+            executors.awaitTermination(480000, MILLISECONDS);
```

Comment:
nit: Should this be somehow related to `maxTestDurationMillis`?

Reply:
....and perhaps at least warn, if not fail outright, if we aren't able to terminate in time, which means the whole test may not have run?

Reply:
Fixed.